src/Evaluator.hs

```diff
@@ -0,0 +1,214 @@
+module Evaluator where
+
+import           ValueSyntax
+import           Syntax
+import           Environment
+import           Utility
+
+import qualified Data.Text                     as T
+
+evalError :: Show a => Expr a -> String -> b
+evalError expr typeString =
+    error
+        $  "Expression"
+        ++ show expr
+        ++ "\ncannot be evaluated to type: "
+        ++ typeString
+
+exprTypeError :: Show a => Expr a -> String -> b
+exprTypeError expr typeString =
+    error $ "Expression" ++ show expr ++ "\nis not of type " ++ typeString
+
+evalPrimExpr :: PrimExpr -> PrimVal
+evalPrimExpr (IntLit   n) = IntVal n
+evalPrimExpr (CharLit  c) = CharVal c
+evalPrimExpr (FloatLit f) = FloatVal f
+evalPrimExpr (BoolLit  b) = BoolVal b
+evalPrimExpr UnitLit      = UnitVal
+
+evalPrimitive :: Show a => Expr a -> Environment -> IO Value
+evalPrimitive (Primitive primExpr _ _) env =
+    return $ PrimitiveVal $ evalPrimExpr primExpr
+evalPrimitive expr _ = exprTypeError expr "Primitive"
```

Comment:
What's the point of this case? It should never happen and it's only there to protect from the programmer errors so it seems to be a bad pattern/design. I think the specific `eval*` functions could just take the unwrapped `Expr a` contents (for example `evalPrimitive :: PrimExpr -> IO Value`).  If the environment is not needed in some cases then it should not be passed. This applies to most of the helper functions defined in this file.